### PR TITLE
Make tray setup work when agents are removed and setup is ran from a different crc binary

### PR DESCRIPTION
**Pull request by anjannath**
**Fixes:** Issue #1140 #1141 

## Solution/Idea

1. When daemon/tray agent is removed using the `launchctl remove` command, `crc setup` was unable to restart it, since the plist file was removed and running `launchctl start` requires the agent to be loaded, therefore we load the agent plist before trying to start it as part of preflight check.
2. When the plist file already exists `crc setup` wouldn't recreate the daemon plist file with the current crc binary's path, so we make the `DaemonPlistFileExists` check to always fail and trigger the recreation of the plist file whenever setup is ran.

## Proposed changes
1. Load plist before starting it
2. Improve `launchd.AgentRunning` function to handle when an agent is removed from launchd
3. Force creation of daemon plist every time `crc setup` is ran

## Testing

1. run `launchctl remove crc.tray; launchctl remove crc.daemon`
2. Now when `crc setup --enable-experimental-fetures` is ran you should have a working tray
3. Download an older version of crc (1.6.0) and enable the tray using `crc setup --enable-experimental-features`
4. Check the crc version from the tray by clicking on _About_
5. Now run `crc setup --enable-experimetal-features` from the crc binary built from this PR.
6. Again check the version like in step 4, and the commit sha from the crc version should match the commit id from HEAD of this patch set. or just check inside `~/Library/LaunchAgents/crc.daemon.plist` that the first string in the `ProgramArguments` key is the path to the crc binary used in step 5.

### CODE REVIEW

File: pkg/crc/preflight/preflight_check_tray_darwin.go [hunk 1]
```diff
@@ -35,13 +35,22 @@ type TrayVersion struct {
...
-		return fmt.Errorf("Daemon plist file does not exist")
-	}
-	return nil
+	// crc setup can be ran from any location in the
+	// users computer, and we need to update the plist
+	// file with the path of the crc binary which was
+	// used to run setup, to force it this check needs
+	// to always fail so the fix routine is triggered
+
+	return fmt.Errorf("Daemon plist file does not exist")
```
Comment (praveenkumar) [inline pkg/crc/preflight/preflight_check_tray_darwin.go:44]: If this check always going to return the `error` then what is the use of this check we should just remove it.

Comment (anjannath) [reply pkg/crc/preflight/preflight_check_tray_darwin.go:44]: Ohh, that'd be nice, this is just there to trigger the fix, can we make the fix run without having the check..?

Comment (praveenkumar) [reply pkg/crc/preflight/preflight_check_tray_darwin.go:44]: @anjannath why not, may be you can change the name without `fix` prefix and run it for every single time?

Comment (anjannath) [reply pkg/crc/preflight/preflight_check_tray_darwin.go:44]: @praveenkumar In that way we have to completely move this out of the preflight checks, now we need this to fill the `preflightCheck` struct, if we leave out the check member in that struct, i think we'll hit a segfault.

File: pkg/os/launchd/launchd_darwin.go [hunk 2]
```diff
@@ -103,7 +104,11 @@ func StopAgent(label string) error {
 
 // RestartAgent restarts a launchd agent
 func RestartAgent(label string) error {
-	err := StopAgent(label)
+	err := LoadPlist(label)
```
Comment (cfergeau) [inline pkg/os/launchd/launchd_darwin.go:107]: Not clear why we need this change, if the agent was already running, then wasn't the plist also loaded already?

File: pkg/os/launchd/launchd_darwin.go [hunk 3]
```diff
@@ -120,8 +125,8 @@ func AgentRunning(label string) bool {
 	if err != nil {
 		return false
 	}
-	if strings.TrimSpace(string(out)) == "-" {
-		return false
+	if match, err := regexp.MatchString(`^\d*$`, strings.TrimSpace(string(out))); err == nil && match {
```
Comment (cfergeau) [inline pkg/os/launchd/launchd_darwin.go:128]: I'd add a small comment such as "match pid". 
`^\d*$` will match an empty string, I don't think we want that if we're looking for a pid.

File: pkg/os/launchd/launchd_darwin.go [hunk 4]
```diff
@@ -7,6 +7,7 @@ import (
 	goos "os"
 	"os/exec"
 	"path/filepath"
+	"regexp"
 	"strings"
```
Comment (cfergeau) [inline pkg/os/launchd/launchd_darwin.go:11]: In the commit log of " Issue #1140 Improve AgentRunning function " can you change "o/p" to "output"? took me a while to expand it :)

File: pkg/crc/preflight/preflight_check_tray_darwin.go [hunk 5]
```diff
@@ -42,6 +42,7 @@ func checkIfDaemonPlistFileExists() error {
 }
 
 func fixDaemonPlistFileExists() error {
```
Comment (cfergeau) [inline pkg/crc/preflight/preflight_check_tray_darwin.go:47]: Commit log is "Issue #1140 Load plist file before trying to start a launchd agent", feel free to detail why this is needed, why this was working before without this, ...

File: pkg/crc/preflight/preflight_check_tray_darwin.go [hunk 6]
```diff
@@ -63,6 +64,7 @@ func checkIfTrayPlistFileExists() error {
 }
 
 func fixTrayPlistFileExists() error {
+	logging.Debug("Creating tray plist file")
```
Comment (cfergeau) [inline pkg/crc/preflight/preflight_check_tray_darwin.go:75]: the addition of debugging statements could go in its own commit.

File: pkg/crc/preflight/preflight_check_tray_darwin.go [hunk 7]
```diff
@@ -79,6 +81,10 @@ func checkIfDaemonAgentRunning() error {
 }
 
 func fixDaemonAgentRunning() error {
+	logging.Debug("Starting daemon agent")
+	if err := launchd.LoadPlist(daemonAgentLabel); err != nil {
+		return err
+	}
 	return launchd.StartAgent(daemonAgentLabel)
```
Comment (cfergeau) [inline pkg/crc/preflight/preflight_check_tray_darwin.go:96]: almost tempted to suggest using a `fixAgentRunnig` helper

File: pkg/crc/preflight/preflight_check_tray_darwin.go (hunk 1 above)
Comment (cfergeau) [reply pkg/crc/preflight/preflight_check_tray_darwin.go:44]: I don't know if it's expensive to recreate this plist file, or if it has side effects we want to avoid, but an alternative could be to change `checkIfDaemonPlistFileExists` to 
- generate a string containing to the content the plist file should have for the current binary
- compare that string with the content of the plist file we have on disk
- only return an error if this does not match
If recreating the plist file is cheap enough, then maybe this is overkill..

File: pkg/os/launchd/launchd_darwin.go (hunk 3 above)
Comment (anjannath) [reply pkg/os/launchd/launchd_darwin.go:128]: Ahh, thanks so much, should be `^\d+$`

File: pkg/crc/preflight/preflight_check_tray_darwin.go (hunk 1 above)
Comment (anjannath) [reply pkg/crc/preflight/preflight_check_tray_darwin.go:44]: i didn't notice an increase in time consumed, so its very cheap to create it, happens near instantly

Comment (cfergeau) [reply pkg/crc/preflight/preflight_check_tray_darwin.go:44]: it's not just file creation, but file creation + LoadPlist + RestartAgent which I was asking if there are noticeable side effects.

File: pkg/os/launchd/launchd_darwin.go (hunk 2 above)
Comment (anjannath) [reply pkg/os/launchd/launchd_darwin.go]: This was a left over from debug/testing, fixed now!

File: pkg/os/launchd/launchd_darwin.go (hunk 4 above)
Comment (anjannath) [reply pkg/os/launchd/launchd_darwin.go:11]: Done!

File: pkg/crc/preflight/preflight_check_tray_darwin.go (hunk 1 above)
Comment (anjannath) [reply pkg/crc/preflight/preflight_check_tray_darwin.go:44]: oh, I did not notice any side effects, seems to work as expected

File: pkg/os/launchd/launchd_darwin.go [hunk 8]
```diff
@@ -113,15 +114,17 @@ func RestartAgent(label string) error {
 // AgentRunning checks if a launchd service is running
 func AgentRunning(label string) bool {
 	// This command return a PID if the process
-	// is running, otherwise returns "-"
+	// is running, otherwise returns "-" and empty
+	// ouput if the agent is not loaded in launchd
```
Comment (cfergeau) [inline pkg/os/launchd/launchd_darwin.go:118]: "ouput" -> "output"
and I think I would say "or empty output" instead of "and .." since we'll either get "-" if loaded but not running, or "" if not loaded.

File: pkg/crc/preflight/preflight_check_tray_darwin.go (hunk 1 above)
Comment (praveenkumar) [reply pkg/crc/preflight/preflight_check_tray_darwin.go:44]: @cfergeau @anjannath I would still think if we are going to keep it then better to update the error message because  `Daemon plist file does not exist` seems confusing to me, what about just put `Ignoring this check` as error message and explanation already part of comments?

File: pkg/crc/preflight/preflight_check_tray_darwin.go [hunk 9]
```diff
@@ -35,13 +35,20 @@ type TrayVersion struct {
...
-		return fmt.Errorf("Daemon plist file does not exist")
-	}
-	return nil
+	// crc setup can be ran from any location in the
+	// users computer, and we need to update the plist
+	// file with the path of the crc binary which was
+	// used to run setup, to force it this check needs
+	// to always fail so the fix routine is triggered
+
+	return fmt.Errorf("Ignoring this check and triggering creation of daemon plist")
```
Comment (praveenkumar) [inline pkg/crc/preflight/preflight_check_tray_darwin.go:44]: 👍
